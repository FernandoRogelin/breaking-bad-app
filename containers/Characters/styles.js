import styled, { css } from 'styled-components';

import ButtonComponent from '#/components/Button';

export const Top = styled.div`
  display: flex;
  padding: 8px 100px 8px 30px;
  justify-content: space-between;
  border-bottom: 1px solid #919191;
`;

export const Name = styled.p`
  margin: 0;
  color: #8c9aa3;
  font-size: 20px;
`;

export const Text = styled.p`
  margin: 0;

  @media only screen and (max-width: 440px) {
    &:last-child {
      display: none;
    }
  }
`;

export const Itens = styled.div`
  display: flex;
  margin: 0 15px;
  padding: 6px 15px;
  align-items: center;
  justify-content: space-between;

  &:not(:last-child) {
    border-bottom: solid 1px #ddd;
  }

  @media only screen and (max-width: 440px) {
    flex-direction: column;
  }
`;

export const Image = styled.div`
  width: 100px;
  height: 100px;
  border-radius: 50%;
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;
  background-image: url(${props => props.background});
`;

export const Title = styled.h1`
  margin: 0;
`;

export const Header = styled.header`
  width: 40%;
  display: flex;
  justify-content: flex-end;
`;

export const Button = styled(ButtonComponent)`
  font-size: 12px;
  padding: 5px 10px;

  &:not(:last-child) {
    margin-right: 10px;
  }
`;

export const Section = styled.section`
  display: flex;
  align-items: center;
  flex-direction: column;
`;
