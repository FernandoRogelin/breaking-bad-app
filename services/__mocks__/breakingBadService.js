export const getCharacters = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-07-1958',
      char_id: 1,
      img:
        'https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg',
      name: 'Walter White',
      nickname: 'Heisenberg',
      occupation: ['High School Chemistry Teacher', 'Meth King Pin'],
      portrayed: 'Bryan Cranston',
      status: 'Presumed dead',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-24-1984',
      char_id: 2,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/f/f2/Jesse_Pinkman2.jpg/220px-Jesse_Pinkman2.jpg',
      name: 'Jesse Pinkman',
      nickname: "Cap n' Cook",
      occupation: ['Meth Dealer'],
      portrayed: 'Aaron Paul',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '08-11-1970',
      char_id: 3,
      img:
        'https://s-i.huffpost.com/gen/1317262/images/o-ANNA-GUNN-facebook.jpg',
      name: 'Skyler White',
      nickname: 'Sky',
      occupation: [
        'House wife',
        'Book Keeper',
        'Car Wash Manager',
        'Taxi Dispatcher',
      ],
      portrayed: 'Anna Gunn',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '07-08-1993',
      char_id: 4,
      img:
        'https://media1.popsugar-assets.com/files/thumbor/WeLUSvbAMS_GL4iELYAUzu7Bpv0/fit-in/1024x1024/filters:format_auto-!!-:strip_icc-!!-/2018/01/12/910/n/1922283/fb758e62b5daf3c9_TCDBRBA_EC011/i/RJ-Mitte-Walter-White-Jr.jpg',
      name: 'Walter White Jr.',
      nickname: 'Flynn',
      occupation: ['Teenager'],
      portrayed: 'RJ Mitte',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: 'Unknown',
      char_id: 5,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/c/c1/Hank_Schrader2.jpg/220px-Hank_Schrader2.jpg',
      name: 'Henry Schrader',
      nickname: 'Hank',
      occupation: ['DEA Agent'],
      portrayed: 'Dean Norris',
      status: 'Deceased',
    },
  ],
});

export const getCharacterDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    appearance: '1,2,3',
    birthday: null,
    char_id: 31,
    img:
      'https://vignette.wikia.nocookie.net/breakingbad/images/c/c9/AdamPinkman.png/revision/latest?cb=20141117180836&path-prefix=es',
    name: 'Adam Pinkman',
    nickname: "Jesse's dad",
    occupation: 'unknown',
    portrayed: 'Michael Bofshever',
    status: 'Alive',
  },
});

export const getEpisodesList = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      air_date: '01-20-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '1',
      episode_id: 1,
      season: '1',
      title: 'Pilot',
    },
    {
      air_date: '01-27-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'White White Jr.',
        'Krazy-8',
      ],
      episode: '2',
      episode_id: 2,
      season: '1',
      title: "Cat's in the Bag...",
    },
    {
      air_date: '02-10-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '3',
      episode_id: 3,
      season: '1',
      title: "...And the Bag's in the River",
    },
    {
      air_date: '02-17-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '4',
      episode_id: 4,
      season: '1',
      title: 'Cancer Man',
    },
  ],
});

export const getEpisodeDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    air_date: '05-03-2009',
    characters:
      'Walter White, Jesse Pinkman, Skyler White, Hank Schrader, Marie Schrader, White White Jr., Saul Goodman, Jane Margolis',
    episode: '9',
    episode_id: 16,
    season: '2',
    title: '4 Days Out',
  },
});

export const getQuotes = jest.fn().mockResolvedValue({
  status: 200,
  data: {},
});
