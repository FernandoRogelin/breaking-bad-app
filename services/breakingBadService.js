import axios from 'axios';

export const urlBase = 'https://www.breakingbadapi.com/api/';

export const getCharacters = () => axios.get(`${urlBase}characters`);

export const getCharacterDetails = id =>
  axios.get(`${urlBase}characters/${id}`);

export const getEpisodesList = () => axios.get(`${urlBase}episodes`);

export const getEpisodeDetails = id => axios.get(`${urlBase}episodes/${id}`);

export const getQuotes = () => axios.get(`${urlBase}quotes`);
