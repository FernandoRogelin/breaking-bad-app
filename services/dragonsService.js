import axios from 'axios';

export const urlBase =
  'http://5c4b2a47aa8ee500142b4887.mockapi.io/api/v1/dragon';

export const getDragons = () => axios.get(urlBase);

export const getDetailsDragon = id => axios.get(`${urlBase}/${id}`);

export const createDragon = ({ name, type, createdAt }) =>
  axios.post(urlBase, {
    name,
    type,
    createdAt,
  });

export const deleteDragon = id => axios.delete(`${urlBase}/${id}`);

export const editDragon = (id, { name, type, createdAt }) =>
  axios.put(`${urlBase}/${id}`, {
    name,
    type,
    createdAt,
  });
