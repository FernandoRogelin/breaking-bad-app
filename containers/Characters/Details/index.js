import React, { useState, useEffect } from 'react';

import { format } from 'date-fns';
import { getCharacterDetails } from '#/services/breakingBadService';

import {
  Wrapper,
  Title,
  Content,
  Text,
  Info,
  Header,
  Button,
  Items,
} from './styles';

export default function Details({ id, setStateDetailsPage }) {
  const [characterDetails, setCharacterDetails] = useState();
  const [fetched, setFetched] = useState(false);

  async function handleCharacterDetails() {
    try {
      const { data } = await getCharacterDetails(id);

      setFetched(true);
      setCharacterDetails(data);
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!fetched) {
      handleCharacterDetails();
    }
  }, [fetched]);

  return (
    <Wrapper>
      <Header>
        <Button onClick={setStateDetailsPage}>Return</Button>
        <Title>Character Details</Title>
      </Header>
      {characterDetails && (
        <Content>
          {characterDetails[0].name && (
            <Items>
              <Text>Name:</Text>
              <Info>{characterDetails[0].name}</Info>
            </Items>
          )}
          {characterDetails[0].portrayed && (
            <Items>
              <Text>Portrayed:</Text>
              <Info>{characterDetails[0].portrayed}</Info>
            </Items>
          )}
          {characterDetails[0].nickname && (
            <Items>
              <Text>Nickname:</Text>
              <Info>{characterDetails[0].nickname}</Info>
            </Items>
          )}
          {characterDetails[0].birthday && (
            <Items>
              <Text>Birthday:</Text>
              <Info>{format(characterDetails[0].birthday, 'MM/DD/YYYY')}</Info>
            </Items>
          )}
          {characterDetails[0].status && (
            <Items>
              <Text>Status:</Text>
              <Info>{characterDetails[0].status}</Info>
            </Items>
          )}
        </Content>
      )}
    </Wrapper>
  );
}
