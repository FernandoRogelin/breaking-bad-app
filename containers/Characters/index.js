import React, { useState, useEffect } from 'react';

import { Link } from '#/server/routes';
import { getCharacters } from '#/services/breakingBadService';
import Content from '#/components/Content';
import Details from './Details';

import {
  Top,
  Name,
  Text,
  Itens,
  Image,
  Title,
  Header,
  Button,
  Section,
} from './styles';

export default function Characters() {
  const [characters, setCharacters] = useState();
  const [isShowDetails, setIsShowDetails] = useState(false);
  const [id, setId] = useState('');
  const [fetched, setFetched] = useState(false);

  async function handleCharactersList() {
    try {
      const { data } = await getCharacters();

      setFetched(true);
      setCharacters(data);
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!fetched) {
      handleCharactersList();
    }
  }, [fetched]);

  function handleDetailsPage(id) {
    setId(id);
    setIsShowDetails(!isShowDetails);
  }

  return (
    <div>
      {isShowDetails ? (
        <Details id={id} setStateDetailsPage={handleDetailsPage} />
      ) : (
        <Section>
          <Title>Characters List</Title>
          <Header>
            <Link route="episodes">
              <Button data-testid="all-episodes-button">All episodes</Button>
            </Link>
            <Link route="quotes">
              <Button data-testid="quotes-button">Quotes</Button>
            </Link>
          </Header>
          {characters && (
            <Content data-testid="character-items">
              <Top>
                <Text>Name</Text>
                <Text>Actions</Text>
              </Top>
              {characters
                .sort((a, b) =>
                  a.name < b.name ? -1 : a.name > b.name ? 1 : 0
                )
                .map(({ name, char_id, img }) => (
                  <Itens key={char_id}>
                    <div>
                      <Image background={img} />
                      <Name>{name}</Name>
                    </div>
                    <Button
                      onClick={() => handleDetailsPage(char_id)}
                      data-testid="details-character-button"
                    >
                      Character details
                    </Button>
                  </Itens>
                ))}
            </Content>
          )}
        </Section>
      )}
    </div>
  );
}
