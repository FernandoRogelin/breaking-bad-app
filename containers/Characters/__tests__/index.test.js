import React from 'react';
import { render, wait, fireEvent } from 'react-testing-library';

import { getCharacters } from '#/services/breakingBadService';
import Characters from '..';

jest.mock('#/services/breakingBadService');

describe('Characters container', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Characters snapshot', () => {
    const { container } = render(<Characters />);

    expect(container).toMatchSnapshot();
  });

  test('Inicial flow', async () => {
    const { getByTestId } = render(<Characters />);

    await wait(() => expect(getCharacters).toHaveBeenCalledTimes(1));

    expect(getByTestId('character-items')).toBeVisible();

    expect(getByTestId('details-character-button')).toBeVisible();
  });

  test('Details page', async () => {
    const { getByTestId } = render(<Characters />);

    await wait(() => expect(getCharacters).toHaveBeenCalledTimes(1));

    fireEvent.click(getByTestId('details-character-button'));
  });
});
