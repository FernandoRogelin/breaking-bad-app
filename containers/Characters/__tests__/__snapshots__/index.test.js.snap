// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters container Characters snapshot 1`] = `
<div>
  <div>
    <section
      class="styles__Section-sc-1ixkgfy-8 gMoorc"
    >
      <h1
        class="styles__Title-sc-1ixkgfy-5 gHHPKM"
      >
        Characters List
      </h1>
      <header
        class="styles__Header-sc-1ixkgfy-6 dGfNBf"
      >
        <button
          class="Button-n3cyj4-0 styles__Button-sc-1ixkgfy-7 faLEsp"
          data-testid="all-episodes-button"
        >
          All episodes
        </button>
        <button
          class="Button-n3cyj4-0 styles__Button-sc-1ixkgfy-7 faLEsp"
          data-testid="quotes-button"
        >
          Quotes
        </button>
      </header>
    </section>
  </div>
</div>
`;
